dev-tools/ls-apis/src/workspaces.rs

```diff
@@ -0,0 +1,270 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+//! Combines information about multiple `Workspace`s
+
+use crate::api_metadata::AllApiMetadata;
+use crate::cargo::Workspace;
+use anyhow::{anyhow, ensure, Context, Result};
+use camino::Utf8PathBuf;
+use cargo_metadata::Package;
+use cargo_metadata::PackageId;
+use std::collections::BTreeMap;
+use std::collections::BTreeSet;
+use std::sync::Arc;
+
+/// Thin wrapper around a list of workspaces that makes it easy to query which
+/// workspace has which package
+pub(crate) struct Workspaces {
+    workspaces: BTreeMap<String, Workspace>,
+}
+
+impl Workspaces {
+    /// Use `cargo metadata` to load workspace metadata for all the workspaces
+    /// that we care about
+    ///
+    /// The data found is validated against `api_metadata`.
+    ///
+    /// On success, returns `(workspaces, warnings)`, where `warnings` is a list
+    /// of potential inconsistencies between API metadata and Cargo metadata.
+    pub fn load(
+        api_metadata: &AllApiMetadata,
+    ) -> Result<(Workspaces, Vec<anyhow::Error>)> {
+        // First, load information about the "omicron" workspace.  This is the
+        // current workspace so we don't need to provide the path to it.
+        let omicron = Arc::new(Workspace::load("omicron", None)?);
+
+        // In order to assemble this metadata, Cargo already has a clone of most
+        // of the other workspaces that we care about.  We'll use those clones
+        // rather than manage our own.
+        //
+        // To find each of these other repos, we'll need to look up a package
+        // that comes from each of these workspaces and look at where its local
+        // manifest file is.
+        //
+        // Loading each workspace involves running `cargo metaata`, which is
```

Comment:
```suggestion
        // Loading each workspace involves running `cargo metadata`, which is
```

I've profiled `cargo metadata` before (since nextest has to invoke it), and it's painfully slow -- both CPU (parsing toml, serializing JSON) and I/O (reading files from disk) intensive. On the other side, deserializing is also quite expensive because the `cargo_metadata` crate's data model is all owned data, not borrowed from the JSON blob.

Reply:
Fixed typo in 684b8b669.